helix-core/src/main/java/org/apache/helix/common/caches/CurrentStateCache.java

```diff
@@ -135,10 +135,15 @@ private void refreshCurrentStatesCache(HelixDataAccessor accessor,
     Set<PropertyKey> cachedKeys = new HashSet<>(_currentStateCache.keySet());
     cachedKeys.retainAll(currentStateKeys);
 
+    List<PropertyKey> reloadKeyList = new ArrayList<>(reloadKeys);
     Map<PropertyKey, CurrentState> newStateCache = Collections.unmodifiableMap(
-        refreshProperties(accessor, new ArrayList<>(reloadKeys), new ArrayList<>(cachedKeys),
+        refreshProperties(accessor, reloadKeyList, new ArrayList<>(cachedKeys),
             _currentStateCache));
 
+    // reloadKeyList is updated by refreshProperties()
+    reloadKeys.clear();
+    reloadKeys.addAll(reloadKeyList);
+
     // if the cache was not initialized, the previous state should not be included in the snapshot
```

Comment:
do we use reloadkeys later in this method?  Why not pass reloadKeyList directly here?

Reply:
reloadKeys are of type Set, which is used later in line 149. reloadKeyList is of type list. Thus need to populate the list to the Set.

Reply:
Then can we change the refreshProperties() to take List?  Convert the type back and forth was actually the reason we had this bug, right? that is why I am thinking whether we can get rid of this conversion at all.

Reply:
This is a change of base class interface, namely refreshProperties() are in the AbstractDataCache class. We can also do it there, though it may change other places that using refreshProperties(). Namely ExternalViewCache,  IdeaStateCache and PropertyCache.

Reply:
These two lines are equal to 
```
reloadKeys = new HashSet<>(reloadKeyList);
```

Another nit comment, can you rename `reloadKeys` to `reloadKeySet` to differentiate from `reloadKeyList`

Reply:
IMHO, refreshProperties needs to be improved. The main issue is that the function takes one parameter as the input and also the output.
Ideally, we shall have 2 separate list/set. One for the input, which keys require reload. The other for the output, read the real reloaded keys.

Reply:
You are right at the point. However, refreshProperties() already return Map<PropertyKey, T> already. It can't return reloadedKeys. I will enhance the comment to stress it will augment the parameter reloadedKeys. Also, I will take Lei's advice to use Set, instead of using List in the refreshProperties()